improc/simulator.py

```diff
@@ -0,0 +1,886 @@
+        and amplifiers and readout electronics.
+
+        Usually this does not change between images
+        from the same survey.
+        """
+        self.sensor = SimSensor()
+
+        self.sensor.bias_mean = self.pars.bias_mean
+        self.sensor.pixel_bias_std = self.pars.bias_std
+        # self.sensor.pixel_bias_map = self.sensor.bias_mean + np.random.poisson(
+        #     self.sensor.pixel_bias_std ** 2, size=self.pars.imsize
+        # ) - self.sensor.pixel_bias_std ** 2
+        self.sensor.pixel_bias_map = np.random.normal(
+            self.sensor.bias_mean,
+            self.sensor.pixel_bias_std,
+            size=self.pars.imsize
+        )
+
+        self.sensor.gain_mean = self.pars.gain_mean
+        self.sensor.pixel_gain_std = self.pars.gain_std
+        self.sensor.pixel_gain_map = np.random.normal(
+            self.sensor.gain_mean,
+            self.sensor.pixel_gain_std,
+            size=self.pars.imsize
+        )
+
+        self.sensor.pixel_qe_std = self.pars.pixel_qe_std
+        self.sensor.pixel_qe_map = np.random.normal(1.0, self.sensor.pixel_qe_std, size=self.pars.imsize)
+        self.sensor.pixel_qe_map[self.sensor.pixel_qe_map < 0] = 0.0
+
+        self.sensor.dark_current = self.pars.dark_current
+        self.sensor.read_noise = self.pars.read_noise
+        self.sensor.saturation_limit = self.pars.saturation_limit
+        self.sensor.bleed_fraction_x = self.pars.bleed_fraction_x
+        self.sensor.bleed_fraction_y = self.pars.bleed_fraction_y
+
+    def make_camera(self):
+        """
+        Generate a camera and save it to the simulator.
+        This includes all the properties of the optics
+        like the optical PSF (not including atmospheric seeing)
+        and the flat field (vignetting).
+        """
+        self.camera = SimCamera()
+        self.camera.vignette_amplitude = self.pars.vignette_amplitude
+        self.camera.vignette_radius = self.pars.vignette_radius
+        self.camera.vignette_offset_x = self.pars.vignette_offset_x
+        self.camera.vignette_offset_y = self.pars.vignette_offset_y
+
+        self.camera.make_vignette(self.pars.imsize)
+
+        self.camera.optic_psf_mode = self.pars.optic_psf_mode
+        self.camera.optic_psf_pars = self.pars.optic_psf_pars
+
+        oversampling = 10
+        self.camera.make_optic_psf(oversampling)  # use a large oversampling just to get the FWHM
+
+    def make_sky(self):
+        """
+        Generate an instance of a sky. This will usually stay the same
+        when taking a series of images at the same pointing.
+        We will assume that if the pointing changes, the sky changes
+        values like seeing and background.
+
+        """
+        self.sky = SimSky()
+        self.sky.background_mean = self.pars.background_mean
+        self.sky.background_std = self.pars.background_std
+        self.sky.background_instance = np.random.normal(self.sky.background_mean, self.sky.background_std)
```

Comment:
Another thing worth doing at some point in the distant future is a variable sky background.  Of course, there's all the "wings of bright stars" stuff we talked about, but just to start some sort of parameterized polynomial.

Reply:
Something like a moon illumination? I think that might actually be relevant to sky flats. I calculate the flats pixel by pixel, but I also start by normalizing each image based on the median sky brightness, so it could make a difference in principle. I will have to think about that.